lindi/LindiH5pyFile/LindiH5pyDataset.py

```diff
@@ -0,0 +1,271 @@
+        external_array_link = zarr_array.attrs.get("_EXTERNAL_ARRAY_LINK", None)
+        if external_array_link and isinstance(external_array_link, dict):
+            link_type = external_array_link.get("link_type", None)
+            if link_type == 'hdf5_dataset':
+                url = external_array_link.get("url", None)
+                name = external_array_link.get("name", None)
+                if url is not None and name is not None:
+                    client = self._get_external_hdf5_client(url)
+                    dataset = client[name]
+                    assert isinstance(dataset, h5py.Dataset)
+                    return dataset[selection]
+        if self._compound_dtype is not None:
+            # Compound dtype
+            # In this case we index into the compound dtype using the name of the field
+            # For example, if the dtype is [('x', 'f4'), ('y', 'f4')], then we can do
+            # dataset['x'][0] to get the first x value
+            assert self._compound_dtype.names is not None
+            if isinstance(selection, str):
+                # Find the index of this field in the compound dtype
+                ind = self._compound_dtype.names.index(selection)
+                # Get the dtype of this field
+                dt = self._compound_dtype[ind]
+                if dt == 'object':
+                    dtype = h5py.Reference
+                else:
+                    dtype = np.dtype(dt)
+                # Return a new object that can be sliced further
+                # It's important that the return type is Any here, because otherwise we get linter problems
+                ret: Any = LindiH5pyDatasetCompoundFieldSelection(
+                    dataset=self, ind=ind, dtype=dtype
+                )
+                return ret
+            else:
+                raise TypeError(
+                    f"Compound dataset {self.name} does not support selection with {selection}"
+                )
+
+        # We use zarr's slicing, except in the case of a scalar dataset
+        if self.ndim == 0:
+            # make sure selection is ()
+            if selection != ():
+                raise TypeError(f'Cannot slice a scalar dataset with {selection}')
+            return zarr_array[0]
+        return zarr_array[selection]
+
+    def _get_external_hdf5_client(self, url: str) -> h5py.File:
+        if url not in _external_hdf5_clients:
+            if url.startswith("http://") or url.startswith("https://"):
+                ff = remfile.File(url)
+            else:
+                ff = open(url, "rb")  # this never gets closed
+            _external_hdf5_clients[url] = h5py.File(ff, "r")
+        return _external_hdf5_clients[url]
+
+
+def _resolve_references(x: Any):
+    if isinstance(x, dict):
+        if '_REFERENCE' in x:
+            return LindiH5pyReference(x['_REFERENCE'])
+        else:
+            for k, v in x.items():
+                x[k] = _resolve_references(v)
```

Comment:
When would `x` be a dict outside of being a special reference dict?

Reply:
It's true that dict attributes are not allowed in hdf5, afaik. I was imagining this could happen for example if you have a dataset where the entries are arbitrary objects, and you have a reference embedded in one of those objects. But maybe that's not possible...

Reply:
I see. I don't think that's possible... and in the converter, hdf5 object array values are only allowed to be bytes/str/Reference. Just to reduce complexity, mind if I change this to an exception?

Reply:
sounds good